Fix handler not disconnected when removing non visible pages using RemovePage()

Vignesh-SF3580 opened the pull request:
<!-- Please let the below note in for people that find this PR -->
> [!NOTE]
> Are you waiting for the changes in this PR to be merged?
> It would be very helpful if you could [test the resulting artifacts](https://github.com/dotnet/maui/wiki/Testing-PR-Builds) from this PR and let us know in a comment if this change resolves your issue. Thank you!

### Issue Details
When a non-visible page is removed from the navigation stack using RemovePage(), its handler is not disconnected.

### Root Cause
The RemovePage() method removes the page from the navigation stack but does not explicitly disconnect its handler.

### Description of Change
The handler is now properly disconnected when a page is removed using RemovePage().

Regression PR: https://github.com/dotnet/maui/pull/24887

### Tested the behavior in the following platforms
 
- [x] Android
- [x] Windows
- [x] iOS
- [x] Mac

### Issues Fixed

Fixes https://github.com/dotnet/maui/issues/32239

### Screenshots

**Android:**
| Before Issue Fix | After Issue Fix |
|----------|----------|
| <video width="300" height="600"  src="https://github.com/user-attachments/assets/890c3e4b-f520-4826-b942-6ec9e04233d1"> | <video width="300" height="600"  src="https://github.com/user-attachments/assets/40ebc025-7298-4b43-8816-4932a01fb563">) |

**iOS:**
| Before Issue Fix | After Issue Fix |
|----------|----------|
| <video width="300" height="600"  src="https://github.com/user-attachments/assets/484e17cd-bd83-415e-af27-99acf893b203"> | <video width="300" height="600"  src="https://github.com/user-attachments/assets/062a1c6d-047c-4bde-add2-d2d261763c87">) |

Vignesh-SF3580 committed:
fixed - disconnect handlers when removing pages using RemovePage

Vignesh-SF3580 committed:
updated the changes

Vignesh-SF3580 committed:
added comments.

Vignesh-SF3580 committed:
added test.

Vignesh-SF3580 committed:
Update Issue32239.cs

Vignesh-SF3580 committed:
updated the fix.

Vignesh-SF3580 committed:
updated

Vignesh-SF3580 commented on src/Controls/src/Core/NavigationPage/NavigationPage.Legacy.cs:
```diff
@@ -252,6 +252,10 @@ void RemovePage(Page page)
 
+			// Only disconnect handlers if the removed page is not visible.
+			// When the current page is removed, the handler is disconnected through PopAsync().
+			page?.DisconnectHandlers();
```
I’ve removed the redundant disconnection logic from NavigationProxy and updated the implementation in ShellSection. This ensures that DisconnectHandler is called only once and avoids any duplicate disconnection.

StephaneDelcroix committed:
Improve comments and fix missing newlines

- Updated comments in NavigationPage.cs and NavigationPage.Legacy.cs to accurately describe the code flow
- Added missing trailing newlines to test files

StephaneDelcroix reviewed (approved):
## ✅ Approved

This PR correctly fixes the handler disconnection issue when removing non-visible pages using `RemovePage()`.

### Summary of Changes

| File | Change |
|------|--------|
| `NavigationPage.cs` | Added `page?.DisconnectHandlers()` in `OnRemovePage` |
| `NavigationPage.Legacy.cs` | Added `page?.DisconnectHandlers()` in `RemovePage` |
| `ShellSection.cs` | Added `page?.DisconnectHandlers()` in `OnRemovePage` |
| `NavigationProxy.cs` | Removed duplicate `DisconnectHandlers()` call |

### Improvements Made

I've pushed a commit that:
1. **Fixed misleading comments** - Updated comments to accurately describe the code flow (this code path is only reached for non-visible pages)
2. **Added missing trailing newlines** - Fixed cosmetic issue in test files

### Verification

- ✅ Fix addresses the root cause from regression PR #24887
- ✅ Proper test coverage with UI automation
- ✅ All four platforms tested by author
- ✅ Null-safe implementation
- ✅ No duplicate disconnection (NavigationProxy call removed)

Great work! 🚀

StephaneDelcroix reviewed (changes requested):
Requesting changes: Adding a unit test in Controls.Core.UnitTests to complement the UI test.

StephaneDelcroix committed:
Add unit tests for handler disconnection on RemovePage

Added two unit tests to verify handler disconnection behavior:
- RemovePageDisconnectsHandlerForNonVisiblePage: Verifies the page is properly removed from navigation stack
- RemovePageCallsDisconnectHandlersOnRemovedPage: Verifies handler is disconnected when page with handler is removed

StephaneDelcroix reviewed (dismissed):
## ✅ Approved

This PR correctly fixes the handler disconnection issue when removing non-visible pages using `RemovePage()`.

### Summary of Changes

| File | Change |
|------|--------|
| `NavigationPage.cs` | Added `page?.DisconnectHandlers()` in `OnRemovePage` |
| `NavigationPage.Legacy.cs` | Added `page?.DisconnectHandlers()` in `RemovePage` |
| `ShellSection.cs` | Added `page?.DisconnectHandlers()` in `OnRemovePage` |
| `NavigationProxy.cs` | Removed duplicate `DisconnectHandlers()` call |

### Improvements Made

I've pushed commits that:
1. **Fixed misleading comments** - Updated comments to accurately describe the code flow
2. **Added missing trailing newlines** - Fixed cosmetic issue in test files
3. **Added unit tests** - New tests in `NavigationUnitTest.cs`:
   - `RemovePageDisconnectsHandlerForNonVisiblePage` - Verifies page is properly removed from navigation stack
   - `RemovePageCallsDisconnectHandlersOnRemovedPage` - Verifies handler is disconnected when page with handler is removed

### Test Coverage

| Test Type | Location | Purpose |
|-----------|----------|---------|
| **Unit Test** | `NavigationUnitTest.cs` | Verifies handler disconnection logic |
| **UI Test** | `Issue32239.cs` | End-to-end validation on device |

### Verification

- ✅ Fix addresses the root cause from regression PR #24887
- ✅ Unit tests pass locally
- ✅ All four platforms tested by author
- ✅ Null-safe implementation
- ✅ No duplicate disconnection (NavigationProxy call removed)

Great work! 🚀

StephaneDelcroix committed:
Remove redundant UI test

The unit tests in NavigationUnitTest.cs adequately verify the handler disconnection behavior. The UI test adds maintenance burden without additional value.

StephaneDelcroix reviewed (dismissed):
## ✅ Approved

Clean, minimal fix for handler disconnection when removing non-visible pages.

### Changes (5 files, +65/-1)

| File | Change |
|------|--------|
| `NavigationPage.cs` | Added `page?.DisconnectHandlers()` in `OnRemovePage` |
| `NavigationPage.Legacy.cs` | Added `page?.DisconnectHandlers()` in `RemovePage` |
| `ShellSection.cs` | Added `page?.DisconnectHandlers()` in `OnRemovePage` |
| `NavigationProxy.cs` | Removed duplicate call (moved to implementations) |
| `NavigationUnitTest.cs` | Added 2 unit tests |

### Unit Tests
- `RemovePageDisconnectsHandlerForNonVisiblePage` - Verifies page removal from stack
- `RemovePageCallsDisconnectHandlersOnRemovedPage` - Verifies handler disconnection

All tests pass. LGTM! 🚀

rmarinho committed:
Improve comments and tests for handler disconnection on RemovePage

- Remove line number references from comments (prone to becoming stale)
- Add handler disconnection verification to existing test
- Rename test for clarity (RemovePageCallsDisconnectHandlersOnRemovedPage -> RemovePageDisconnectsHandlerForRemovedRootPage)
- Add Shell-specific test for RemovePage handler disconnection

rmarinho reviewed (dismissed):
## ✅ Approved

This PR properly fixes the handler disconnection issue when using `RemovePage()` on non-visible pages.

### Summary of Changes
- **Root cause fix**: Moves `DisconnectHandlers()` from `NavigationProxy` to the concrete implementations (`NavigationPage`, `NavigationPage.Legacy`, `ShellSection`)
- **No duplicate disconnection**: By removing from the proxy and adding to implementations, handlers are disconnected exactly once
- **Clean comments**: Comments explain the behavior without brittle line number references

### Test Coverage
- ✅ `RemovePageDisconnectsHandlerForNonVisiblePage` - Tests removing middle page from 3-page stack
- ✅ `RemovePageDisconnectsHandlerForRemovedRootPage` - Tests removing root page when another page is on top
- ✅ `RemovePageDisconnectsHandlerInShell` - Tests Shell navigation scenario

All tests pass for both `useMaui: true` and `useMaui: false` scenarios, covering both the modern and legacy navigation implementations.

LGTM! 🚀

PureWeen commented:
/azp run

PureWeen committed:
Add agent review session for PR #32289
